Feature: Creating Virtual Devices on non-Windows systems

lhpt2 opened the pull request:
As stated above I included the functions to create and delete virtual input and output ports into the library.

I wrapped these functions in the `PortMidi` struct implementation and added a mutex with a vector to hold the
created virtual devices. This way all virtual devices automatically get freed when the `PortMidi` struct drops (added logic in the drop trait implementation) and it is easily possible to keep an overview of all virtual devices (requesting a list of virtual devices and the number of virtual devices also possible now).

I also added an example and some test calls to the `test_main` function.

I hope the code is not too messy. I'm relatively new to Rust and it might be that I didn't always find the most efficient solution :'D

Commit:
Add Virtual Device functions to Rust-C interface (functions.rs)

Commit:
Add virtual device list to main context

Commit:
Adjust error msg on dropping/deleting virtual devices

Commit:
Add virtual_device_count() function to context.rs

Commit:
Test virtual_device_count() == 0 in test_main

Commit:
Mutex for virtual device list and create_virtual_output() function

Commit:
Doc annotation for virtual_device_count() function

Commit:
Add Error handling for invalid device name in create_virtual_output()

Commit:
Add virtual_devices() method for getting info about all created virtual devices

Commit:
Refactor create_virtual_output() into create_virtual_device() and add create_virtual_input()

Commit:
Add delete_virtual_device() function and refactor drop() function for PortMidi

Commit:
Docstring for delete_virtual_device()

Commit:
Virtual Devices example program added

Commit:
Add tests for virtual devices to test_main

musitdev commented:
Hi @lhpt2, thanks for this addition. I try to review it asap.
Add a comment if I haven't done it by the end of next week.

lhpt2 committed:
Change create_virtual_input() and create_virtual_output() to take &str instead of String

musitdev commented on src/context.rs:
```diff
@@ -1,14 +1,20 @@
 /// Initializes PortMidi on creation and terminates it on drop.
 pub struct PortMidi {
     device_count: u32,
+    virtual_devs: Mutex<Vec<PortMidiDeviceId>>
```
I don't see why you use a Mutex? If it's to make the PortMidi Struct send and sync, the best is to let the user decide and wrap the context in a Mutex or anything he prefers.

lhpt2 replied:
I use a mutex for making the `virtual_devices.rs` example work. 
If I get you right you suggest making the user wrap the PortMidi struct inside an `Arc<Mutex<>>` in a whole for using it inside different threads? 

Wouldn't that make access to the non-mutable methods unnecessarily complex? In every case, I have to make sure the `virtual_devs` member and the corresponding methods are not accessed by different threads at the same time as they are mutable.

musitdev replied:
I said that because PortMidi struct can be used in a multi thread env, async or mono thread. It depends on the user needs, not the lib one. For example, why you use a Mutex and not a RwLock? It depends on how the lib is used. If you define a mutex everyone will have to manage this mutex even if there's no need. There's a cost at using a mutex.
If you think that locking the mutex is complex, define your own struct and hide the locking call in it.
Why you want interior mutability? create_virtual_device can use a self mutable, it's even more logic that you mutate thing when creating a device. Interior mutability should only be use when you want to hide an internal algo that need mutability on a function that shouldn't be mutable like a counter on a get.
In my opinion, the fact that you keep the id of the created virutal device is not part of the lib, but more how the developer use the lib. If I create only one device and don't need to keep it. You should remove the virtual_devs ids vector and only provide PortMidi function.
Last thought, what you should add to PortMidi lib is to call Pm_DeleteVirtualDevice when the return Id is dropped. It allows easing the management of the devices.

musitdev reviewed (commented):
I think you should remove the mutex.
There are some formatting issues. Try to pass cargo fmt.
For the rest it's ok for me.

lhpt2 committed:
Reformatting with cargo fmt

lhpt2 committed:
Remove virtual_devs list (managing virtual devices will be handed to the user of the lib), deleting will be added to drop of DeviceInfo struct (if virtual)

lhpt2 committed:
Alternative approach using an own struct that deletes the virtual device on drop
Couldn't decide if I just saved id inside or wrapped it around a DeviceInfo struct.

I went with a DeviceInfo struct in the end and returned a reference from devinfo() function
to make clear that the VirtualDevice struct is not allowed to go out of scope or else
the virtual device is freed.

Of course this would need a copy in the rest of the API functions...

lhpt2 committed:
Refactor vdevice.rs to make it less tempting dropping VirtualDevice instantly after creation (removing devinfo method)

lhpt2 commented:
So I created an alternative with a dedicated VirtualDevice struct now, which frees the virtual device on drop.

I wasn't sure if I should save only the id inside or a whole DeviceInfo struct. I went with the DeviceInfo struct in the end to make gathering info about the device easier, not sure if this is what you thought of.

The thing I'm a little concerned about now is that it isn't really clear to the developer, that he has to keep the VirtualDevices he created or else they will be instantly dropped again but I guess this is just something to be mentioned in the documentation thus.

lhpt2 committed:
Run cargo fmt

musitdev reviewed (approved):
Ok it's fine for me. I think this way give more flexibility.

musitdev commented:
Merged and publish with the version 0.3.0
